contracts/semi-bridge/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+extern crate alloc;
+
+use ink_lang as ink;
+
+#[ink::contract(env = pink_extension::PinkEnvironment)]
+mod semi_bridge {
+    use alloc::{string::String, string::ToString, vec::Vec};
+    use index::prelude::*;
+    use index::utils::ToArray;
+    use ink_storage::traits::{PackedLayout, SpreadLayout};
+    use pink_web3::ethabi::{Bytes, Uint};
+    use pink_web3::keys::pink::KeyPair;
+    use pink_web3::signing::Key;
+    use primitive_types::{H160, H256, U256};
+    use scale::{Decode, Encode};
+
+    #[ink(storage)]
+    pub struct SemiBridge {
+        owner: AccountId,
+        key: [u8; 32],
+        config: Option<Config>,
+    }
+
+    #[derive(Encode, Decode, Debug, PackedLayout, SpreadLayout)]
+    #[cfg_attr(
+        feature = "std",
+        derive(scale_info::TypeInfo, ink_storage::traits::StorageLayout)
+    )]
+    struct Config {
+        rpc: String,
+        bridge_address: [u8; 20],
+    }
+
+    #[derive(Encode, Decode, Debug, PartialEq, Eq)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        BadOrigin,
+        NotConfigurated,
+        KeyRetired,
+        KeyNotRetiredYet,
+        UpstreamFailed,
+        BadAbi,
+        FailedToGetStorage,
+        FailedToDecodeStorage,
+        FailedToEstimateGas,
+        FailedToCreateExecutor,
+    }
+
+    type Result<T> = core::result::Result<T, Error>;
+
+    impl SemiBridge {
+        #[ink(constructor)]
+        pub fn default() -> Self {
+            Self {
+                owner: Self::env().caller(),
+                config: None,
+                key: Self::key_pair().private_key(),
+            }
+        }
+
+        /// Configures the bridge
+        #[ink(message)]
+        pub fn config(&mut self, rpc: String, bridge_address: H160) -> Result<()> {
+            self.ensure_owner()?;
+            self.config = Some(Config {
+                rpc,
+                bridge_address: bridge_address.into(),
+            });
+            Ok(())
+        }
+
+        /// Import a private key to override the interior account
+        #[ink(message)]
+        pub fn set_account(&mut self, private_key: Vec<u8>) -> H160 {
+            self.key = private_key.to_array();
+            self.wallet()
+        }
+
+        /// Returns the wallet address
+        #[ink(message)]
+        pub fn wallet(&self) -> H160 {
+            let keypair: KeyPair = self.key.into();
+            keypair.address()
+        }
+
+        /// Returns BadOrigin error if the caller is not the owner
+        fn ensure_owner(&self) -> Result<()> {
+            if self.env().caller() == self.owner {
+                Ok(())
+            } else {
+                Err(Error::BadOrigin)
+            }
+        }
+
+        /// Derives the key pair on the fly
+        fn key_pair() -> pink_web3::keys::pink::KeyPair {
```

Comment:
As the above comment, we will have a dedicated account manager do account deriving. But it's ok if this PR is just to demonstrate